error reporting

cc10512 opened the pull request:
This PR addresses the issue of repeated error and warning messages when
compiler passes are repeated, by introducing error/warning types, and
an error catalog. Error types classify the errors and impose a format
that allows the compiler to automatically filter repeated
messages. The filtering is based on the type of error and the source
code location of the object that reports the error. Thus, it allows
multiple error types per source code line, and ensures that only one
error is reported even if the message is raised multiple times. Error
codes and formats are defined in `lib/error_catalog.[h,cpp]`. Backends
can extend the codes and formats as needed (and they are encouraged to
do so).

To ease the transition to typed errors and warnings, free form
messages whose first argument (`%1%`) is an `IR::Node` (or more
precisely a class that implements `Util::IHasSourceInfo` interface),
will be converted automatically to typed errors that use the format
argument of the message rather than the error catalog predefined
formats.

We converted the errors in the bmv2 backend to typed errors, as well 
as a number of errors and warnings in the frontend. The conversion is
by no means finalized, however, the added support for transition does
handle a fair number of the duplicates. And I would like to get feedback
from others on what is their view before continuing to convert.

Also introduces the `FATAL_ERROR` macro to print a message and raise
an CompilationError exception.

And finally adds documentation on how to use to the coding standards.

**For reviewers**
I broke the PR in several commits to ease the review process:
- the first commit (https://github.com/p4lang/p4c/commit/fb2c8bcd01515f27ff0f11699fa63c20af0788b6) fixes a couple of warnings that point to potential issues with scoping
- the second commit (https://github.com/p4lang/p4c/commit/5f8b3c17b11e177f7e9fbbf5ed4c8578a943819d) adds the infrastructure support for typed errors. 
- and finally the https://github.com/p4lang/p4c/commit/7ac417bef01beca36e1e67304d2c7d6be64a2e29 converts bmv2 errors and https://github.com/p4lang/p4c/commit/9bcf80709b41cde3201821ecba9cc99f27d5e20e converts frontend errors and warnings.
It might help if you review these separately. 

**Still to do**
The sample outputs will need to be adjusted. Will do that after initial feedback.

mbudiu-vmw commented:
Can you point to an example which had redundant errors?
Some passes are run only once. Is it worth doing this transformation for such passes?

mbudiu-vmw commented on backends/bmv2/common/control.cpp:
```diff
@@ -93,10 +93,10 @@ void ControlConverter::convertTableEntries(const IR::P4Table *table,
                 }
             } else {
-                ::error("unkown key match type '%1%' for key %2%", matchType, k);
+                ::error(ErrorType::ERR_UNKNOWN, k, "key match type " + matchType);
```
Whenever you converted something to a concatenation you will get a different result; in particular you won't information about the source line position of the construct which is appended. Sometimes this may be desired.

cc10512 replied:
The line number is always printed for the first argument (k in this case).

mbudiu-vmw commented on backends/bmv2/common/action.cpp:
```diff
@@ -129,7 +129,7 @@ void ActionConverter::convertActionBody(const IR::Vector<IR::StatOrDecl>* body,
             }
         }
-        ::error("%1%: not yet supported on this target", s);
+        ::error(ErrorType::ERR_EXPRESSION, s, "not yet supported on this target");
```
I find these errors harder to understand and write for the developer.
Wouldn't it be easier to add an enum to the error function describing the category?

cc10512 replied:
That's what ERR_EXPRESSION is (well, not really an enum because of C++ limitations on extending enum classes, and thus forcing me to use a class with static const members). In addition to the enum, this proposal structures the error so that we have a program object that we can use as source. If you prefer the original messages (typos included for some of them! :-) ) we can keep the messages and simply use the format from here rather than the error catalog.

mbudiu-vmw commented on docs/CodingStandardPhilosophy.md:
```diff
@@ -77,3 +79,100 @@ repeating (visual) patterns should usually be avoided.  This is an aspect
+meaning that helps extensibility and maintainability by a large group
+of developers.
+
+## Handling errors
```
This is too detailed for the coding standards. The longer this section, the less likely it is for anyone to read it.
Also, we should keep this very high level; these are really implementation details and not principles.

cc10512 replied:
Where do we put the details?

mbudiu-vmw replied:
We don't have a good place. The readme, where you removed some details from, isn't great either.
Some of this is also in the compiler-design powerpoint, but you can't find that using search.
People will have to read the code anyway to understand how to report errors, so perhaps this is implied.

mbudiu-vmw commented on lib/CMakeLists.txt:
```diff
@@ -17,6 +17,7 @@ set (LIBP4CTOOLKIT_SRCS
 	compile_context.cpp
 	crash.cpp
 	cstring.cpp
+        error_catalog.cpp
```
indentation?

cc10512 commented:
Here is an example of the reduction in the number of warnings:
Using the bmv2 backend on p4c/testdata/p4_14_samples/switch_20160512/switch.p4
```
wc -l switch_trim.log switch.log
      38 switch_trim.log
      43 switch.log
```
The difference is:
```
diff switch_trim.log switch.log
23a24,25
> warning: 1024: Size specified for table without keys
> warning: 1024: Size specified for table without keys
38a41,43
> warning: Action selector '.ecmp_action_profile' is never referenced by a table and cannot be included in bmv2 JSON
> warning: Action selector '.fabric_lag_action_profile' is never referenced by a table and cannot be included in bmv2 JSON
> warning: Action selector '.lag_action_profile' is never referenced by a table and cannot be included in bmv2 JSON
```
The last three are filtered out because the errors are not typed, however, it is likely that you would like both warnings. Specifying a type allows you to issue different warnings for the same line.
The two logs are attached.
[switch_trim.log](https://github.com/p4lang/p4c/files/2713657/switch_trim.log)
[switch.log](https://github.com/p4lang/p4c/files/2713658/switch.log)

For an internal version of a switch profile, the reduction is quite more dramatic roughly 3x:
```
wc -l switch_MSDC_trim.log switch_MSDC.log
     145 switch_MSDC_trim.log
     494 switch_MSDC.log
```
Such a gap is the difference between people paying attention and fixing the warnings vs. continuously ignoring them.

mbudiu-vmw commented:
The first 2 errors are really a bug; they should display the table name as well, and probably they would not be repeated. This is tracked by https://github.com/p4lang/p4c/issues/1504.

So in this example I don't see any errors that are really repeated.
The fact that you lost the last 3 warning is also a bug.

Today most errors will prevent the next pass from running, so only warnings may be repeated.
Perhaps you could do the changes first just for warnings - this is likely the biggest payout.

hanw commented on lib/error_catalog.h:
```diff
@@ -0,0 +1,53 @@
+
+/// enumerate supported errors
+/// Backends should extend this class with additional errors in the range 500-999 and
+/// warnings in the range 1500-2141.
```
wWhy is the range important? What happens if user decides to lower certain error to warning at compiler invocation time? Does the encoding move from range 500-999 to 1500-2141?

cc10512 replied:
No, the value of the enum won't move. The change in behavior is at the point of handling the warning/error, while the value just determines the format in the catalog.
The ranges are not essential, but I thought it would be good to partition the space for ease of further developing the notion of an error catalog.

hanw commented on lib/error_catalog.cpp:
```diff
@@ -0,0 +1,37 @@
+#include "error_reporter.h"
+
+std::map<int, const std::string> ErrorReporter::errorCatalog = {
+    { ErrorType::ERR_UNKNOWN,     "%1%: Unknown %2%"},      // args: %1% - IR node, %2% - what
```
I was hoping each catalog can be a complete sentence about the error and then developer can provide more reasoning on the specific occurrence of the error. The error message we print should be like a git commit message, with a title and a body. Title is mostly fixed template by the compiler, body is provided by the developer.

For example:

```
Error: "file.p4", line xxx, column xxx: (error code)
  Unsupported primitive clone_egress_to_egress
  The supported primitives are clone_ingress_to_egress, clone_ingress_to_ingress.
```

mbudiu-vmw replied:
Han's proposal would also make it easier to localize the errors if we decide to do it.

cc10512 replied:
I think this is reasonable. One potential issue is that we end up with too many errors codes, which will be confusing for developers to pick. So some amount of customization for the title will be needed as well. That said, I agree with giving it a try once we hear from a few more folks. 

I hope that it is clear what problem we're trying to address. Additional suggestions on how to make this work for both users and developers are more than welcome.

antoninbas commented on lib/error_catalog.cpp:
```diff
@@ -0,0 +1,37 @@
+#include "error_catalog.h"
+#include "error_reporter.h"
+
+std::map<int, const std::string> ErrorReporter::errorCatalog = {
```
Why is this declared in error_reporter.h and defined in error_catalog.cpp? Circular dependency?

I hate global variables like this as you don't know when they are constructed. I'd rather see a singleton pattern here (with a method that can be called by backends to register new error codes in the catalog?)

cc10512 commented:
I've addressed the following reviewer comments:
- make the ErrorCatalog a singleton class and defined the format and title in error_catalog.cpp
- added methods to extract the error name and error format
- added support for naming the warnings so that classes of warnings can be disabled on the command line
- added types for most warnings (and a few errors)
- fixed the "bugs" that remove additional warnings on the same object. Note that there are several places where warnings do not get a source code object. Those were left un-typified to avoid eager filtering.
- updated the reference outputs.

I've tried Han's suggestion of having a "title" for the warnings and unfortunately there is no good way of covering entire classes with a single title -- it has to be customized, and thus, it will require the specialization that I implemented the first time. Since you found that hard to write messages for, I went for the simpler approach of simply printing the name of the "offending" object.

Please take a look and let me know what you think.

mbudiu-vmw commented on testdata/p4_16_samples_outputs/copy-frontend.p4:
```diff
@@ -6,7 +6,7 @@ control c(inout bit<32> b) {
     S s2_0;
     @name("c.a") action a() {
-        s2_0 = { 32w0 };
+        s2_0 = {32w0};
```
there should be no whitespace difference here. is this because you are missing a rebase?

cc10512 replied:
yes.

mbudiu-vmw commented on testdata/p4_16_samples_outputs/decl.p4-stderr:
```diff
@@ -1,10 +1,10 @@
-decl.p4(33): warning: y shadows y
+decl.p4(33): [--Wwarn=shadow] warning: y: y shadows y
```
this error message has become worse.

cc10512 replied:
Agreed. Removed the duplication.

mbudiu-vmw commented on testdata/p4_16_samples_outputs/default-switch.p4-stderr:
```diff
@@ -1,4 +1,4 @@
-default-switch.p4(13): warning: SwitchCase: label following default label
+default-switch.p4(13): [--Wwarn=ordering] warning: SwitchCase: SwitchCase: label following default label
```
there are a whole bunch of errors from now on where there is a duplicated IR node name.

cc10512 replied:
fixed.

cc10512 commented on backends/bmv2/common/expression.cpp:
```diff
@@ -433,7 +432,7 @@ void ExpressionConverter::postorder(const IR::Mux* expression)  {
     mapExpression(expression, result);
     if (simpleExpressionsOnly) {
-        ::error("%1%: expression to complex for this target", expression);
+        ::error(ErrorType::ERR_EXPRESSION, "too complex for this target", expression);
```
Added UNSUPPORTED_ON_TARGET error type.

cc10512 commented on testdata/p4_16_errors_outputs/action-bind1.p4-stderr:
```diff
@@ -1,4 +1,4 @@
-action-bind1.p4(22): error: a: Parameter b must be bound
+action-bind1.p4(22): [--Werror=legacy] error:  a: Parameter b must be bound
```
Fixed.

cc10512 commented on testdata/p4_16_errors_outputs/issue306.p4-stderr:
```diff
@@ -1,36 +1,6 @@
-issue306.p4(30): error: Structure struct my_packet does not have a field stack
+issue306.p4(30): [--Werror=legacy] error:  Structure struct my_packet does not have a field stack
```
Legacy errors are errors that have not been typified. I added a TYPE_ERROR to address this issue.

cc10512 commented on testdata/p4_16_errors_outputs/stack3.p4-stderr:
```diff
@@ -1,30 +1,6 @@
-stack3.p4(34): error: header h[]: Size of header stack type should be a constant
+stack3.p4(34): [--Werror=legacy] error:  header h[]: Size of header stack type should be a constant
```
Added TYPE_ERROR.

cc10512 commented on testdata/p4_16_errors_outputs/twovarbit.p4-stderr:
```diff
@@ -1,10 +1,4 @@
-    varbit<120> u;
-                ^
-twovarbit.p4(2): error: x and u: multiple varbit fields in a header
+twovarbit.p4(2): [--Werror=legacy] error:  x and u: multiple varbit fields in a header
```
It is printed in the latest.

cc10512 commented:
@mbudiu-vmw I believe I addressed all your concerns. Can you please take another look?

mbudiu-vmw commented on lib/error_catalog.cpp:
```diff
@@ -0,0 +1,98 @@
+#include "error_catalog.h"
+
+// -------- Errors -------------
+const int ErrorType::LEGACY_ERROR      =   0;
```
Why can't these be just an enum?
Do we care about their concrete values outside of the compiler?

cc10512 replied:
Per the class comment:
```
/// enumerate supported errors
/// Backends should extend this class with additional errors in the range 500-999 and
/// warnings in the range 1500-2141.
/// It is a class and not an enum class because in C++11 you can't extend an enum class
class ErrorType {
```

mbudiu-vmw commented on testdata/p4_16_errors_outputs/issue306.p4-stderr:
```diff
@@ -1,36 +1,6 @@
-issue306.p4(30): error: Structure struct my_packet does not have a field stack
+issue306.p4(30): [--Werror=type-error] error: Structure struct my_packet does not have a field stack
```
It looks like for this error the two arguments should be swapped, because these are actually not duplicate messages.

cc10512 replied:
Attached is how the errors look if I switch it. It is true that it identifies all uses of the .stack member, but it does seem a bit redundant? Let me know if you prefer it this way.
[issue306.p4-stderr.txt](https://github.com/p4lang/p4c/files/2829218/issue306.p4-stderr.txt)

mbudiu-vmw replied:
I think it's better. We could avoid printing the struct position altogether, although it's probably useful the first time.
